### ISSUE TITLE
print's flush argument is missing from builtins, I'd like to add it

### ISSUE THREAD

**Issue by mbdevpl** [python/typeshed#36]
In Python 3.3 and later, I can write this code:

```python
"""
    Printing ten dots one by one in a single line.
"""

import time

if __name__ == '__main__':
    for _ in range(10):
        print('.', end='', flush=True)
        time.sleep(1 / 10)
    print()
```

docs: https://docs.python.org/3/library/functions.html#print

I'm familiar with doing pull requests, and I'd like to add information about new print signature to typeshed. However, before I do, I'd like to learn how to contribute. Currently there is this file: `/builtins/3/builtins.pyi`

So from what I understand, I would be creating `/builtins/3.3/builtins.pyi` and in there I would add:

```python
def print(*values: Any, sep: str = ' ', end: str = '\n', file: IO[str] = None, flush: bool=False) -> None: ...
```

in place of the old one:

```python
def print(*values: Any, sep: str = ' ', end: str = '\n', file: IO[str] = None) -> None: ...
```

... but what about everything else? Should I add all other updates to `builtins.pyi` between 3 and 3.3 before I request a pull?

**Comment by mbdevpl** [comment]
Just for completeness, here's my output from `mypy printing.py`:
```
printing.py:9: error: Unexpected keyword argument "flush" for "print"
...\Python35\Lib\mypy\typeshed\builtins\3\builtins.pyi:673: note: "print" defined here
```

**Comment by JukkaL** [comment]
I think that any new 3.3-only additions to builtins should go to `/builtins/3/builtins.pyi`. It's worth adding a comment about things that aren't included in 3.2, though.

In the longer term we could have conditional definitions in the stubs that define a different signature depending on the current Python version but mypy doesn't support this yet.

**Comment by gvanrossum** [comment]
It would be a shame to have to duplicate builtins.py just because print is
different in 3.3+. As long as mypy doesn't support conditional definitions,
would this work?

# 3/builtins.pyi:
from _print_function import print as print

# 3.2/_print_function.pyi:
def print(...)  # old version

# 3.3/_print_function.pyi:
def print(...)  # with flush= added

IIUC the module search path should do the right thing then.

**Comment by mbdevpl** [comment]
I've read through `/mypy/build.py` (in mypy project) and line 254 states:
```python
    # E.g. for Python 3.5, try 2and3/, then 3/, then 3.5/, then 3.4/, 3.3/, ...
```
Does this make sense? This way new signatures will never be noticed if there are any signatures at all in `3`...

I've tested my quick & dirty solution (duplication with minor change), and in my case (cpython 3.5) the new signature doesn't get picked up at all... `3` has precedence over `3.3`. I had to overwrite the `builtins.pyi` residing in `3` for mypy to pick it up.

@gvanrossum I'd like to test your proposition, but I am doubtful whether the signature will be picked up in case of python 3.0 and 3.1, when `print()` is defined in 3.2.

lines 252..253 in above-mentioned `build.py`:
```python
    versions = ["%d.%d" % (pyversion[0], minor)
                for minor in reversed(range(pyversion[1] + 1))]
```
... `pyversion` being e.g. `[3, 1]` for python 3.1.

The workaround would be to put the old `print()` in `3.0` directory. I will try that now. I won't even try puting it in `3` because then the old signature would be picked even in `3.5`, which is exactly the problem I want to solve...

**Comment by mbdevpl** [comment]
#### `_print_function.pyi` approach

see https://github.com/python/typeshed/compare/master...mbdevpl:print_fix1

The solution partially works. It works great for versions 3.3+, but for earlier there is a strange `None` instead of `"print"`.

Example outputs:

```
>python ...\Python35\Scripts\mypy --python-version 3.0 printing.py
printing.py:9: error: Unexpected keyword argument "flush"
...\Python35\Lib\mypy\typeshed\builtins\3.0\_print_function.pyi:1: note: None defined here
>
```

```
>python ...\Python35\Scripts\mypy --python-version 3.2 printing.py
printing.py:9: error: Unexpected keyword argument "flush"
...\Python35\Lib\mypy\typeshed\builtins\3.0\_print_function.pyi:1: note: None defined here
>
```

```
>python ...\Python35\Scripts\mypy --python-version 3.3 printing.py
>
```

```
>python ...\Python35\Scripts\mypy --python-version 3.4 printing.py
>
```

#### copy&paste approach

see https://github.com/python/typeshed/compare/master...mbdevpl:print_fix2

This seems to work flawlessly:

Example outputs:

```
>python ...\Python35\Scripts\mypy --python-version 3.0 printing.py
printing.py:9: error: Unexpected keyword argument "flush" for "print"
...\Python35\Lib\mypy\typeshed\builtins\3.0\builtins.pyi:673: note: "print" defined here
>
```

```
>python ...\Python35\Scripts\mypy --python-version 3.2 printing.py
printing.py:9: error: Unexpected keyword argument "flush" for "print"
...\Python35\Lib\mypy\typeshed\builtins\3.0\builtins.pyi:673: note: "print" defined here
>
```

```
>python ...\Python35\Scripts\mypy --python-version 3.3 printing.py
>
```

```
>python ...\Python35\Scripts\mypy --python-version 3.4 printing.py
>
```

Any comments on these results?

**Comment by mbdevpl** [comment]
#### `_print_function.pyi` + voodoo programming approach

see https://github.com/python/typeshed/compare/master...mbdevpl:print_fix3

The solution partially works. It works great for versions 3.3+, but for earlier there is a strange `None` instead of `"print"`.

Example outputs:

```
>python ...\Python35\Scripts\mypy --python-version 3.0 printing.py
printing.py:9: error: Unexpected keyword argument "flush"
...\Python35\Lib\mypy\typeshed\builtins\3.0\_print_function.pyi:1: note: None defined here
>
```

```
>python ...\Python35\Scripts\mypy --python-version 3.2 printing.py
printing.py:9: error: Unexpected keyword argument "flush"
...\Python35\Lib\mypy\typeshed\builtins\3.0\_print_function.pyi:1: note: None defined here
>
```

```
>python ...\Python35\Scripts\mypy --python-version 3.3 printing.py
>
```

```
>python ...\Python35\Scripts\mypy --python-version 3.4 printing.py
>
```

**Comment by gvanrossum** [comment]
It looks fine. I wonder if the 'None' in the error message (the 'note:'
part) isn't a bug in mypy. I'll merge your PR once I've confirmed that.

**Comment by gvanrossum** [comment]
So as far as I can tell, the problem is that when you re-export print, mypy creates a new variable named print and copies the type, but doesn't consider it to be a "function definition" so it doesn't have a name attribute. I can fix this with a little hack in the code where the message is printed, but I need to consult with Jukka whether that's the right solution.

Also, PEP 484 says that in a stub file, using the "import ... as ..." syntax, as well as "from ... import ... as ...", should export the name from the stub, but it doesn't seem to do this (as you've found out). I've filed #1111 for this. It's possible that a fix for that might solve the 'None' issue more elegantly, so I'm holding off on merging your PR.

**Comment by mbdevpl** [comment]
Roger. In the meantime, the pull request is hanging out here: https://github.com/python/typeshed/pull/41 - Let me know when any further action is needed.

**Comment by matthiaskramm** [comment]
``` # E.g. for Python 3.5, try 2and3/, then 3/, then 3.5/, then 3.4/, 3.3/, ...```

You have a point. This is weird.

It should look into the most language-specific directory first, and then fall back to more generic locations.

I'll do a mypy PR.

**Comment by gvanrossum** [comment]
Fixed well enough by 0938e8f5de7d0bfbe62c0d492df276746a41e059.
